docs/Tutorial_DAMP.ipynb

```diff
@@ -0,0 +1,378 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #24.        For the given index i, segmentize the array np.arange(i) into 
</pre><p>Is there a better way to describe the task of this function?</p><p> </p><p>"Given the index <code>i</code>, divide the array <code>np.arange(i)</code> into chunks. Ensure the last chunk's size is <code>chunksize_init</code>, with the preceding chunks doubling in size. The output consists of (start, stop) indices of each chunk."</p><p> </p><p><strong>Also, we may add: </strong></p><p>The left-most chunk always start at 0.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/920/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Reply:
<p>This function seems weird. It's not clear why you need to end with the last element being <code>chunksize_init</code> </p><p> </p><p>Something feels backwards here. </p>

Reply:
<p>Also, do we actually need to call the function multiple times? Or is it possible to call the function once and then shift the indices somehow?</p>

Reply:
> Something feels backwards here

Well, it is backward (I mean... the process itself is backward). but, I trust your lens! Please allow me to think more and see if I can come up with something that is more elegant.


> Also, do we actually need to call the function multiple times? Or is it possible to call the function once and then shift the indices somehow?

I wanted to do that but I decided to start simple. What we can do is to keep shifting the start, stop indices of all chunks by one in each iteration. So, for instance, if I get the chunks for the subsequences in `T[:idx]`. Then, I can find the chunks for `T[:(idx+1)]` by just adding start/stop indices of chunks by one to the right. We just need to keep track of the left-most chunk though as it can become the power two of a number at some point. In such case, the number of chunks will be increased by one.

Reply:
[FYR...before I forget]
Another approach is to just offset from the original stop index `idx` by `s`, `2s`, `4s`, `8s`, and so on, where `s` is power of two.  The good thing is that the difference between any two consecutive offset is STILL a power of two. This may result in cleaner code.  IIRC, I think I saw something similar in the MATLAB code before. Going to check that as well.

Reply:
The MATLAB code uses the following lines for updating the start / stop of each chunk:

```
# Query is at index `i`, i.e. `query = T(i:i+SubsequenceLength-1)`
# X is the chunk size, initially set to `2 ^ nextpow2(8 * SubsequenceLength)`

X_start = i - X + 1 + (expansion_num * SubsequenceLength)
X_end = i - (X / 2) + (expansion_num * SubsequenceLength)

# and then
approximate_distance = min( real(MASS_V2(T(X_start:X_end), query))); 
```

The term `(expansion_num * SubsequenceLength)` is to take into account the elements of the last subsequence in a chunk that were not considered in the previous chunk. To keep the length of chunk untouched, the `X_start` will have the same shift. 

---

Note 1:
According to the paper / MATLAB code, the length of chunk, `T`, in `core.mass(Q, T)` should be a power-of-two, and NOT the number of subsequences in the chunk.

Note 2:
The reason behind considering the length `power-of-two` for chunk is to speed up the MASS algorithm (according to the authors)

So, I did a quick check by considering different sizes for chunk in range `(2 ** 20 - 1000 - 1, 2 ** 20 + 1000 + 1)`

```
seed = 0
np.random.seed(seed)

T = np.random.rand(10_000_000)
m = 50

T, M_T, Σ_T, T_subseq_isconstant = stumpy.core.preprocess(T, m)

query_idx = 600000
t_lst = []
for stop in range(2 ** 20 - 1000 - 1, 2 ** 20 + 1000 + 1):
    time_start = time.time()

    QT = core.sliding_dot_product(
        T[query_idx : query_idx + m], 
        T[start : stop],
    )
    
    D = core._mass(
        T[query_idx : query_idx + m],
        T[start : stop],
        QT=QT,
        μ_Q=M_T[query_idx],
        σ_Q=Σ_T[query_idx],
        M_T=M_T[start : stop - m + 1],
        Σ_T=Σ_T[start : stop - m + 1],
        Q_subseq_isconstant=T_subseq_isconstant[query_idx],
        T_subseq_isconstant=T_subseq_isconstant[start : stop - m + 1],
    )

    time_end = time.time()
    t_lst.append(time_end - time_start)
```

And, the I plot it:


```
indices = np.arange(2 ** 20 - 1000 - 1, 2 ** 20 + 1000 + 1)
indices = indices[2:]
t_lst = t_lst[2:]

idx = np.flatnonzero(indices == 2 ** 20)[0]

plt.figure(figsize=(20, 5))
plt.scatter(indices[idx], t_lst[idx], color='r', label='chunksize = 2 ** 20')
plt.plot(indices[idx-200 : idx+200], t_lst[idx-200:idx+200]) 
plt.ylabel('running time')
plt.legend()
plt.show()
```

<img width="975" alt="image" src="https://github.com/TDAmeritrade/stumpy/assets/38519522/401f469c-6eac-47c2-910b-7cec71e5a68a">

Well, it seems the running time for the chunk size `2 ** 20` is low (not necessarily the lowest) but it should be okay. To remain faithful to the paper, I am going to consider the length power-of-two for the chunk size.